# MATH 210 Introduction to Mathematical Computing

## February 28, 2018

1. Error formula for Simpson's rule
2. Numerical differentiation

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import scipy.integrate as spi
%matplotlib inline

## 1. Error formula for Simpson's rule

[Simpson's rule](https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule) is an approximation of the definite integral

$$
\int_a^b f(x) \, dx \approx \frac{h}{3} \sum_{k=1}^{N/2} \left( f(x_{2k-2}) + 4f(x_{2k-1}) + f(x_{2k}) \right)
$$

where $N$ is an even number giving the number of subintervals of $[a,b]$, $h = (b - a)/N$ and $x_k = a + k h$.

The natural question to ask is: how good of an approximation is $S_N(f)$?

**Theorem** The error in Simpson's rule is

$$
E_N^S(f) = \left| \int_a^b f(x) \, dx - S_N(f) \right| \leq \frac{(b-a)^5}{180N^4} M
$$

where $|f''''(x)| \leq M$ for all $x \in [a,b]$.

**Example.** Find a value $N$ which guarantees that $E_N^S(f) \leq 0.0001$ for

$$
\int_1^2 \frac{1}{x} dx
$$

Let's compute $f''''(x)$:

\begin{align*}
f(x) &= \frac{1}{x} \\
f'(x) &= -\frac{1}{x^2} \\
f''(x) &= \frac{2}{x^3} \\
f'''(x) &= -\frac{6}{x^4} \\
f''''(x) &= \frac{24}{x^5} \\
\end{align*}

Therefore $|f''''(x)|\leq 24$ for all $x \in [1,2]$ and so

\begin{align*}
E_N^S(f) = \frac{1}{180N^4} 24 &\leq 0.0001 \\
\frac{20000}{15N^4} &\leq 1 \\
\left( \frac{20000}{15} \right)^{1/4} &\leq N \\
\end{align*}

In [2]:
(20000/15)**0.25

6.042750794713537

Therefore $N=8$ (the smallest even integer greater than 6.04) guarantees that $E_N^S(f) \leq 0.0001$.

In [3]:
N = 8
x = np.linspace(1,2,N+1)
y = 1/x
I = spi.simps(y,x)
np.abs(np.log(2) - I) <= 0.0001

True

In [4]:
print(I)

0.693154530655


## 2. Numerical differentiation

There are 3 main [difference formulas](https://en.wikipedia.org/wiki/Finite_difference#Forward,_backward,_and_central_differences) for computing derivatives. The forward difference formula with step size $h$ is an approximation of $f'(a)$ given by

$$
f'(a) \approx \frac{f(a + h) - f(a)}{h}
$$

The backwards difference formula with step size $h$ is an approximation of $f'(a)$ given by

$$
f'(a) \approx \frac{f(a) - f(a - h)}{h}
$$

The central difference formula with step size $h$ is an approximation of $f'(a)$ given by the average of the forward and backwards difference formulas

$$
f'(a) \approx \frac{1}{2} \left( \frac{f(a + h) - f(a)}{h} + \frac{f(a) - f(a - h)}{h} \right) = \frac{f(a + h) - f(a - h)}{2h}
$$

Let's write a function called `central_diff` which takes input parameters `f`, `a` and `h` (with default value $h=0.01$) and returns the central difference formula for $f'(a)$ with step size $h$.

In [5]:
def central_diff(f,a,h=0.01):
    '''Compute the central difference formula f'(a) \approx (f(a+h) - f(a-h))/2h 
    
    Parameters
    ----------
    f : function of one variable
    a : number
    h : number
    
    Returns
    -------
    Central difference formula f(a+h) - f(a-h))/2h.
    '''
    return (f(a + h) - f(a - h))/(2*h)

In [6]:
central_diff(np.cos,0)

0.0

In [7]:
central_diff(np.exp,0,h=0.0001)

1.0000000016668897

We can create small anonymous functions using `lambda` expressions. These are useful for defining one-line functions inline. See the [Python documentation](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions).

In [8]:
central_diff(lambda x : 1/x,1,h=10e-10)

-1.000000082740371

In [9]:
central_diff(lambda x : 1/np.sin(x),np.pi/2,h=10e-5)

0.0

Natural questions: how good of an approximation are forward/backwards and central difference formulae? The errors formulas come from the Taylor error formula:

**Theorem** The Taylor polynomial for $f(x)$ at $x=a$ of degree $n$ with remainder term is given by

$$
f(x) = f(a) + f'(a)(x - a) + \frac{f''(a)}{2}(x-a)^2 + \cdots + \frac{f^{(n)}(a)}{n!}(x-a)^n + \frac{f^{(n+1)}(c)}{(n+1)!}(x-a)^{n+1}
$$

for some $c$ between $x$ and $a$.

We'll derive the error formulas for the difference formulas next time.